prediction_market_agent/agents/custom_agent.py

```diff
@@ -0,0 +1,124 @@
+import json
+import requests
+from typing import Optional
+from prediction_market_agent import utils
+from prediction_market_agent.agents.abstract import AbstractAgent
+from prediction_market_agent.ai_models.openai_ai_models import ChatOpenAIModel
+from prediction_market_agent.tools.google_search import google_search
+from prediction_market_agent.tools.web_scrape_structured import (
+    web_scrape_structured_and_summarized,
+)
+from prediction_market_agent.tools.tool_exception_handler import tool_exception_handler
+
+
+class CustomAgent(AbstractAgent):
+    def __init__(self, verbose: bool = True, max_cycles: int = 10):
+        """
+        Initialize the CustomAgent.
+
+        Args:
+            verbose: Whether to display intermediate thinking steps.
+            max_cycles: The maximum number of thinking steps for the agent.
+        """
+
+        self.verbose = verbose
+        self.max_cycles = max_cycles
+        self.keys = utils.get_keys()
+        self.tools = {
+            "GoogleSearchTool": google_search,
+            "WebScrapingTool": tool_exception_handler(
+                map_exception_to_output={
+                    requests.exceptions.HTTPError: "Couldn't reach the URL. Don't scrape the same domain again."
+                }
+            )(web_scrape_structured_and_summarized),
+        }
+        self.system_prompt = f"""You are a research assistant meant to answer questions placed on a predictive market.
+
+To do that, you can't rely on your training data, but you have access to the following tools in order to answer question:
+
+1. GoogleSearchTool: Given a query, returns a list of URLs. Accepts arguments as a dictionary, with a single key "query".
+2. WebScrapingTool: Given an objective and url, returns the summarized content of the page. In the objective, ask closed-ended questions. Accepts arguments as a dictionary, with keys "objective" and "url".
+
+If you want to use the tool, simply return completion in form of a dictionary with the keys "tool_name" and "tool_params".
```

Comment:
This actually works well?! I'm surprised you are getting good function args with this. Have you seen that you can pass in `tools` as an arg to `client.chat.completions.create`? (see https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools). Any reason you chose to stick them in the system prompt yourself instead of using that?

Reply:
> This actually works well?!

Yeah, from my experience, GPT4 is quite good at following such requests.

> Any reason you chose to stick them in the system prompt yourself instead of using that?

I wanted to check out what will happen. If it will really work. Nice thing is that this is universal, we can try it out with other (open) models, too, without being constrained by OpenAI's format.

In the end the `tools` argument to OpenAI will convert it (somehow) to prompts too, but I'd expect from my implementation to have worse performance, due to different training.  But seems to be okay so far.

Reply:
Makes sense. Yeah I guess performance depends a lot on how the system prompt compares with what's used in the training data. Cool that it works well!